### PR TITLE
Trim managedFields in pod informer

### PR DESCRIPTION
**Pull request by linxiulei**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Trim unnecessary fields in pod cache in kube-scheduler to improve memory efficiency. 

#### Special notes for your reviewer:

Also see prior work #118455. This change is safe because:

* scheduler doesn't use Extract model on Pod, which needs managedFields
* scheduler doesn't use Update on Pod either, which updates `managedFields` if it's not `nil`
* even if scheduler use Update method on Pod in the future, updating Pod with managedFields being `nil` has no impact as `kube-apiserver` calculates the managed fields from the live object

#### Does this PR introduce a user-facing change?

```release-note
Improve memory usage of kube-scheduler by dropping the `.metadata.managedFields` field that kube-scheduler doesn't require.
```

### CODE REVIEW

**Comment by alculquicondor** [comment]
Tentatively
/ok-to-test

cc @apelisse for any gotchas.
We currently only have one use of SSA, but I think it would work as we build the patch from scratch. cc @mimowo as implementor.

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -498,5 +499,14 @@ func newPodInformer(cs clientset.Interface, resyncPeriod time.Duration) cache.Sh
 	tweakListOptions := func(options *metav1.ListOptions) {
 		options.FieldSelector = selector
 	}
-	return coreinformers.NewFilteredPodInformer(cs, metav1.NamespaceAll, resyncPeriod, cache.Indexers{}, tweakListOptions)
+	informer := coreinformers.NewFilteredPodInformer(cs, metav1.NamespaceAll, resyncPeriod, cache.Indexers{}, tweakListOptions)
```
Comment (mimowo) [inline pkg/scheduler/scheduler.go:502]: IIUC it might be problematic if `ExtractPod` function was used to prepare the SSA. Currently we use SSA, but built from scratch: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/preemption/preemption.go#L365-L378.

Still, it seems beneficial to expose this piece of code and explain its consequences. For example, add a comment that this would cause issues if `ExtractPod` was used to prepare the SSA apply based on the pod. Then, there is a chance someone trying to use `ExtractPod` to use SSA finds the comment :)

Comment (alculquicondor) [reply pkg/scheduler/scheduler.go:502]: I think there is a bigger risk.

We currently update the Pending condition with Update. Would that strip the managedFields data in apiserver?
Should we first migrate any Update calls in kube-scheduler into SSA before we can proceed with this change?

Comment (linxiulei) [reply pkg/scheduler/scheduler.go:502]: > I think there is a bigger risk.
> 
> We currently update the Pending condition with Update. Would that strip the managedFields data in apiserver? Should we first migrate any Update calls in kube-scheduler into SSA before we can proceed with this change?

I might be totally wrong, but to share my understandings

If we use Update already, then 

* the caller thinks it's the truth of source to override that object (otherwise any other writers will be affected, does resourceVersion help though?)
* it doesn't care managedManaged (otherwise it will use Patch with SSA option)


Also, even for Update request, it's the apiserver that applies managedFields and apiserver always has the copy to reconcile the stripping caused by Update, right?

Comment (linxiulei) [reply pkg/scheduler/scheduler.go:502]: > IIUC it might be problematic if ExtractPod function was used to prepare the SSA. Currently we use SSA, but built from scratch:

Please correct me if I'm wrong. The ExtractPod doesn't rely on any managedField so stripping those has no impact on ExtractPod. Am I missing anything problematic?

Comment (alculquicondor) [reply pkg/scheduler/scheduler.go:502]: > Please correct me if I'm wrong. The ExtractPod doesn't rely on any managedField so stripping those has no impact on ExtractPod. Am I missing anything problematic?

ExtractPod relies on managedField, but we don't use it in kube-scheduler, so this is not a concern.

The usage of Update seems to be a concern, depending on what apiserver does when receiving an Update with empty managedField

Comment (linxiulei) [reply pkg/scheduler/scheduler.go:502]: According to https://github.com/kubernetes/kubernetes/pull/119556#discussion_r1275306322, the Update will be similar to Create (ie. the exact object with corresponding managedField)

Comment (mimowo) [reply pkg/scheduler/scheduler.go:502]: > The ExtractPod doesn't rely on any managedField so stripping those has no impact on ExtractPod

IIUC and played it with before, it uses managed fields to only populate the fields in the extracted object which match the field manager. Take a look here: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/applyconfigurations/core/v1/pod.go#L50-L63. Also, https://github.com/kubernetes/kubernetes/blob/d4fde1e92a83cb533ae63b3abe9d49f08efb7a2f/staging/src/k8s.io/client-go/applyconfigurations/core/v1/pod.go#L74.

Still, as @alculquicondor mentioned, it is not currently used, still I would like to have some comment that it may impact the uses of SSA which operate on the extracted pods. I'm pretty sure someone will want to use SSA in this fashion and it will be puzzling why it does not work. Leaving a comment can help to find the source of complications. 

cc @apelisse in case I'm missing something.

**Comment by alculquicondor** [comment]
Can you summarize the learnings about why this is safe in the PR description?

/lgtm
/approve

**Comment by linxiulei** [comment]
> Can you summarize the learnings about why this is safe in the PR description?

Done. Thanks!

**Comment by alculquicondor** [comment]
I would argue this is user facing and needs a release note.

/remove-kind cleanup
/kind feature

/release-note-edit
```release-note
Improve memory usage of kube-scheduler by dropping the `.metadata.managedFields` field that kube-scheduler doesn't require.
```

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Improve memory usage of kube-scheduler by dropping the `.metadata.managedFields` field that kube-scheduler doesn't require.
```

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (apelisse) [reply pkg/scheduler/scheduler.go:502]: Thanks, I think you're right. I've personally never been a big fan of the `Extract` workflow, and I don't think I've seen it being used anywhere yet as far as I know, ideally it should continue to be unused. I'm fine with this as it is, feel free to add a comment.

Comment (linxiulei) [reply pkg/scheduler/scheduler.go:502]: Thanks for all the insights. I added a comment to warn the future use of `Extract`. Additionally, #119672 would prevent future use of `Extract`.

**Comment by liggitt** [comment]
A data race on resync was just reported in use of WithTransform in https://github.com/kubernetes/client-go/issues/1348 that may affect this use
